### ISSUE TITLE
troubleshoot mingw on conan-lzma

### ISSUE THREAD

**Issue by solvingj** [bincrafters/community#42]
Initial attempt at adding MinGW to conan-lzma failed: 

https://ci.appveyor.com/project/BinCrafters/conan-lzma/build/1.0.4/job/wfckab1smt798qlq

```
lzma/5.2.3@bincrafters/testing: Copying sources to build folder
lzma/5.2.3@bincrafters/testing: Generator txt created conanbuildinfo.txt
lzma/5.2.3@bincrafters/testing: Calling build()
'.' is not recognized as an internal or external command,
operable program or batch file.
lzma/5.2.3@bincrafters/testing: 
lzma/5.2.3@bincrafters/testing: 
```

Just need someone to try to fix at next opportunity

**Comment by dimi309** [comment]
Am I right to assume that this configuration

`Environment: MINGW_CONFIGURATIONS=4.9@x86_64@seh@posix, 4.9@x86_64@sjlj@posix, 4.9@x86@sjlj@posix, 4.9@x86@dwarf2@posix, 6.3@x86_64@seh@posix, 7.1@x86_64@seh@posix`

... which is used for the mingw build is "plain vanilla" mingw and not using msys? If that is the case it probably cannot work because the conanfile is running configure and make. Since we are building on appveyor, perhaps the only solution might be those mingw or cygwin installers @sigmoidal and I think also @SSE4 were looking into.

This is also related to our conversation about these *nix-on-windows (for lack of a better term) tools: https://github.com/bincrafters/community/issues/13. Note that by now, conan also supports distinguishing between them via subsettings: https://github.com/conan-io/conan/issues/2120. Has anyone used this feature yet? Also, have we reached a common understanding about how to handle this kind of build? For example do we use a cygwin or mingw installer or do we presuppose the presence of cygwin or mingw? I have been doing the latter for the most part (considering it a more light-weight approach) but I suppose some of you prefer the former. Maybe we should try to conclude this discussion and put some guidance, indicating what we have decided, into our templates at some point...

**Comment by SSE4** [comment]
@dimi309 @solvingj @sigmoidal 
I've checked in branch mingw/5.2.3 into conan-lzma, seems to be working locally

**Comment by SSE4** [comment]
@dimi309 to answer your questions:
- usually you wouldn't be able to use plain vanialla MinGW without MSYS or Cygwin, as you have specified due to the configure and make. in general, you will need to have some kind of unix-style environment and tools for majority of builds, unless it's something very trivial.
-  subsettings are used to distinguish target environment. e.g. in MSYS you can produce binaries which may run only in MSYS itself, or in regular Windows. same for Cygwin. majority of users won't need binaries running only in Cygwin/MSYS environment, so it's safe to ignore em, unless you're building build tools (like ninja, nasm, etc).
- we usually presuppose the presence of MinGW and MSYS for build. assumption is that some conan profile is being used which has MinGW, MSYS/Cygwin as build requirements.

**Comment by dimi309** [comment]
@SSE4 I suppose it will work locally if you have msys in your path. The point is that on Appveyor I think it didn't exactly because there was no msys. Thank you for discussing the rest of the points. Here are my ideas on each one, in corresponding order:

- I use plain vanilla mingw extensively, working on projects that use ogg, vorbis, bzip2, zlib, portaudio, freetype, libpng and a bunch of other libraries that get built in that mode, even via conan. So I would not discard it, firstly because many conan packages already support it and also because build times are much faster than with msys and it is indeed the most native way to use gcc on windows as far as I know (unless MS is up to some new tricks again with Visual Studio 😄 ). Anything that uses a cmake configuration or a carefully written make script will work there just fine.
- As you can tell, in my mind it is Cygwin and msys that are the poor relatives in the *nix-on-win toolgroup and I would much rather ignore them altogether (I would assume that you would disagree 😆 ) So, going down that road, which I am afraid we won't be doing though, I would of course totally agree that subsettings would offer us no added value 😄 
- Theoretically maybe there is the presupposition of msys presence but, as I've mentioned in the first point, there are already a lot of conan packages that run without msys. Also, for users like myself who do work on packages with vanilla mingw makefiles, the presence of msys say in a CI environment is far from benign because it breaks cmake -G"MinGW Makefiles" configurations (cmake won't accept to find msys in the path in that case and will crash on purpose).

Naturally, I mention all of these concerns in good faith, without the intention to just convince everyone to do things my way. But I do hope that we will standardise on a common strategy that will cover the needs so much of users that work with msys and cygwin, as of those like myself who work with mingw-light, so to speak.

**Comment by SSE4** [comment]
@dimi309 
- actually, appveyor has MSYS, MinGW and Cygwin installed, see https://www.appveyor.com/docs/build-environment/#mingw-msys-cygwin
- but we don't use them (at least recipes I am aware of), instead mingw_installer and msys_installer packages are used in build.py (passed to conan package tools, then passed to the profile)

**Comment by dimi309** [comment]
But it looks like they are not available because of the way the build is setup for this package on appveyor (I never assumed that appveyor doesn't have an msys option at all). The [failed build](https://ci.appveyor.com/project/BinCrafters/conan-lzma/build/1.0.4/job/wfckab1smt798qlq) ends like this:

```
lzma/5.2.3@bincrafters/testing: Calling build()
'.' is not recognized as an internal or external command,
operable program or batch file.
lzma/5.2.3@bincrafters/testing: 
lzma/5.2.3@bincrafters/testing: ERROR: Package 'a125fd3185659d4b7baf056fdcf10cca690f5491' build failed
lzma/5.2.3@bincrafters/testing: WARN: Build folder C:\Users\appveyor\.conan\data\lzma\5.2.3\bincrafters\testing\build\a125fd3185659d4b7baf056fdcf10cca690f5491
ERROR: lzma/5.2.3@bincrafters/testing: Error in build() method, line 76
	self.build_configure()
while calling 'build_configure', line 68
	env_build.configure(args=args)
	ConanException: Error 1 while executing ./configure --disable-xz --disable-xzdec --disable-lzmadec --disable-lzmainfo --disable-scripts --disable-doc --prefix=C:\Users\appveyor\.conan\data\lzma\5.2.3\bincrafters\testing\build\a125fd3185659d4b7baf056fdcf10cca690f5491\lzma-install --disable-static --enable-shared 
Error while executing:
	 conan create  lzma/5.2.3@bincrafters/testing --profile c:\users\appveyor\appdata\local\temp\1\tmplgnsqzconan_package_tools_profiles\profile 
```

It looks to me like it cannot find `configure`. So I suppose that either the appveyor setup has to change to include msys, or an installer has to be used for the build. Then of course it will never work with vanilla mingw-make, unless subsettings are used for each case. The contrary is true on the other hand. If a package works with vanilla mingw-make it probably also works for msys and cygwin without subsettings (and even on linux and on macos without many changes) which is why I prioritise it and why I have made the points above.

**Comment by SSE4** [comment]
@dimi309 
loader finds configure, it just can't execute it directly (like `./configure` in your case)
in *nix, loader always reads shebang (line with something like `#!/bin/bash`) and executes interpreter automatically. it's not the case for Windows, where you need to execute interpreter explicitly, like:
`bash -l -c "./configure"`
or, use some sort of helper for this, like **tools.run_in_windows_bash**
also, **AutoToolsBuildEnvironment** has argument **win_bash** which will automatically forces conan to invoke everything in bash
you also either have to specify full path to bash.exe executable, or add it into the PATH
conan's helper **tools.run_in_windows_bash** reads environment variable **CONAN_BASH_PATH**, so you can set it
 
to summarize, changes needed:
- pass **win_bash** flag to the **AutoToolsBuildEnvironment**
- set environment variable **CONAN_BASH_PATH** to some sort of bash
- add **bash.exe** to the **PATH** environment variable
that's basically it
if you can take a look at mentioned **mingw/5.2.3** branch, you may ensure that it does exactly that I have specified above

**Comment by dimi309** [comment]
Right, it seems to work now. Thank you! And apologies... I forgot about needing to call bash to execute such commands in Windows.
I was trying to note the importance of using mingw without msys whenever possible but since I've misdiagnosed what was wrong with this particular package my whole point is off-topic for this thread 😄  
I don't know if you would like me to do anything else here or if you can just merge and close this issue.

**Comment by SSE4** [comment]
well, if you know how to build withoy msys, it would be nice to try

**Comment by dimi309** [comment]
I have just taken a look inside the library's distribution archive, at the file xz-5.2.3/windows/INSTALL-MinGW.txt. It says:

> MSYS is always needed to use the GNU Autotools based build system.

... and indeed it only provides instructions for autotools. Moreover, there is no cmake support and no make file that doesn't need a configure command to execute first so, sadly, it seems that it is not possible in this case. I still do think that we should do it before everything else when it is possible though (there I go again 😄 ).

**Comment by SSE4** [comment]
if it's possible then why not - having less dependencies is always better
but since it's not the case for lzma, discussion is not relevant here
so let's continue with reviewing mingw/5.2.3 branch, testing it and possibly setting up some sort of CI for MinGW on appveyor

**Comment by dimi309** [comment]
Good, good.
